### PR TITLE
bug/4878 validate application name

### PR DESCRIPTION
**Pull request by ppittle**
- [x] Needs #171 need to merge first

image

*Issue #, if available:*
- #124 
- #123 
- DOTNET-4878

*Description of changes:*
 sanitize cloud application name recommendations and also validate user input

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by philasmar — APPROVED** [review]

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            // make sure the recommendation doesn't exist already in existingApplications
+            var recommendation = recommendedPrefix;
+            var suffix = 1;
+            while (suffix < 100)
+            {
+                if (existingApplications.All(x => x.StackName != recommendation) && IsValidName(recommendation))
+                    return recommendation;
+
+                recommendation = $"{recommendedPrefix}{suffix++}";
+            }
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:77]: We should use something like GUID which is unique but shorter in length. Perhaps first two segments of GUID or last segment should be good enough for our use case.

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            _directoryManager = directoryManager;
+        }
+
+        public string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications)
+        {
+            // generate recommendation
+            var recommendedPrefix = "deployment";
+
+            if (_fileManager.Exists(target.ProjectPath))
+                recommendedPrefix = Path.GetFileNameWithoutExtension(target.ProjectPath);
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:52]: Customer can have a directory starting with - (dash) or number which isn't accepted by CloudFormation.

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        /// </summary>
+        private readonly Regex _validatorRegex = new ("^[a-zA-Z][a-zA-Z0-9-]{2,126}$", RegexOptions.Compiled);
+
+        public CloudApplicationNameGenerator(IFileManager fileManager, IDirectoryManager directoryManager)
+        {
+            _fileManager = fileManager;
+            _directoryManager = directoryManager;
+        }
+
+        public string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications)
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:46]: There is no check on the length of the application name.
```
The stack name is an identifier that helps you find a particular stack from a list of stacks. A stack name can contain only alphanumeric characters (case-sensitive) and hyphens. It must start with an alphabetic character and can't be longer than 128 characters.
```

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html

I think the generated name should match `^[a-zA-Z][-a-zA-Z0-9]{1,128}$`, I'm not saying to use regex.

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 3 above)
Comment (ganeshnj) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:46]: We can do the same validation on customer entered application for early feedback than trying to deploy and fail.

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 2 above)
Comment (ppittle) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:52]: You are correct.  Down below the `recommendedPrefix` is [sanitized](https://github.com/aws/aws-dotnet-deploy/pull/169/files#diff-a0db7211bc877b00e7c00a810aa3a6a1459080bdc78592fed5541b10b404759bR66) to remove invalid characters at the start of the recommendation.


Here's the test case covering it: https://github.com/aws/aws-dotnet-deploy/pull/169/files#diff-a0db7211bc877b00e7c00a810aa3a6a1459080bdc78592fed5541b10b404759bR66

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 3 above)
Comment (ppittle) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:46]: 1.  Length: `_validatorRegex` does check length: `{2,126}`  I was a little more restrictive than the regex you mentioned which uses `{1,128}`.
2. Early feedback.  Feedback and validation is done as the user enters/selects an Application Name.  They can't even see recommendations until a valid application name is selected.  See the screenshot in the PR Description.

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 1 above)
Comment (ppittle) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:77]: Are you recommending this?

| Current | Guid Proposal |
| --- | --- |
| Select the AWS stack: <li>WebApp</li><li>WebApp1</li><li>WebApp2</li>| Select the AWS stack: <li>WebApp</li><li>WebApp-de8bab2b</li><li>WebApp-e8da26de</li> |

I think I'd rather stick with the numbers.  I find it a bit more approachable and it has a natural sort order.

Do you want to make the case for the Guid Proposal?

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 3 above)
Comment (ppittle) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:46]: Initial regex was a bit too conservative.  Expanded to allow Names are short as 1 character and as long as 128

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 1 above)
Comment (ganeshnj) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:77]: Can we check this with team just to have second opinion?

Comment (normj) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:77]: I'm with PJ on this. This is a user facing name that the user might have to type in. Nobody wants to type in a guid. It also follows the pattern of new projects creating in VS.

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            // sanitize recommendation
+            recommendedPrefix =
+                new string(
+                    recommendedPrefix
+                        .ToCharArray()
+                        .SkipWhile(c => !char.IsLetter(c) && ((int)c) < 127)
+                        .Where(c =>
+                            char.IsNumber(c) ||
+                            char.IsLetter(c) && ((int)c) < 127 ||
+                            c == '-')
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:65]: this makes me more comfortable
`
'a' <= c <= 'z' || 'A' <= c <= 'Z' || '0' <= c <= '9' || c == '-'
`

**Review by ganeshnj — APPROVED** [review]

File: src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs (hunk 4 above)
Comment (ppittle) [reply src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs:65]: interesting suggestion, but think I'm going to leave it as is

**Pull request merged**